### PR TITLE
Initial snowflake destination

### PR DESCRIPTION
**Pull request by jsantos17**
There's some technical debt I'd like to address in a separate PR, since @beckyconning tells me we need to get this out of the door ASAP. So TODO:

1. Figure out where to stage  files. Right now we default to the user stage, but I'm unsure if this is correct.
2. More tests. I need to split `SnowflakeDestination` to make it more testable, especially query generation.

### CODE REVIEW

**Comment by jsantos17** [comment]
[ch9591]

**Comment by jsantos17** [comment]
Another note:

This will never drop a table. I thought about it, but it seems dangerous. We may delete user's data. This has the caveat where pushing a sample, verifying it works and pushing the entire result will not work if the user doesn't drop the sample table first.

File: core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    errs.map(_.show).intercalate(", ")
+
+  // we need to retrieve a SnowflakeConnection from the Hikari transactor.
+  // this is the recommended way to access Snowflake-specific methods
+  // https://docs.snowflake.net/manuals/user-guide/jdbc-using.html#unwrapping-snowflake-specific-classes
+  private def snowflakeConnection(xa: Transactor[F]): Resource[F, SnowflakeConnection] =
+    xa.connect(xa.kernel).map(_.unwrap(classOf[SnowflakeConnection]))
+
+  private def ensureSingleSegment(r: ResourcePath): F[FileName] =
+    r.fold(
```
Comment (wemrysi) [inline core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala]: :golf: 
```scala
r match {
  case file /: ResourcePath.root() => file.pure[F]
  case _ => ...
}
```

File: core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      .query[Unit].stream.transact(xa).compile.drain
+
+  private def doPush(
+    path: ResourcePath,
+    columns: List[TableColumn],
+    bytes: Stream[F, Byte],
+    freshName: String)
+      : Stream[F, Unit] =
+    for {
+      inputStream <- (io.toInputStream[F]: Pipe[F, Byte, InputStream])(bytes)
```
Comment (wemrysi) [inline core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala]: :golf: `bytes.through(io.toInputStream)`

**Comment by beckyconning** [comment]
We should drop the table. I know it seems dangerous but without offset tracking its the only way to update the tables and its the way the other destinations work.

File: core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+      fileName <- Stream.eval(ensureSingleSegment(path))
+
+      _ <- debug(s"Finished staging to file: @~/${fileName.value}")
```
[the remark this answers is not in the document]
Comment (jsantos17) [reply core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala]: Absolutely 🤦‍♂️

Fixed.

File: core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala (hunk 1 above)
Comment (jsantos17) [reply core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala]: That's cool extractor. Will implement!

File: core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+      if (compressed)
+        stagedFile + ".gz"
+      else
+        stagedFile
+
+    fr"COPY INTO" ++
+      Fragment.const(escapeString(tableName)) ++
+      fr0" FROM @~/" ++
+      Fragment.const(fileToLoad) ++ // no risk of injection here since this is fresh name
+      fr"file_format = (type = csv, skip_header = 1)"
```
Comment (beckyconning) [inline core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala]: `(type = 'CSV' field_delimiter = ',' skip_header=1 FIELD_OPTIONALLY_ENCLOSED_BY='"')`

Comment (jsantos17) [reply core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala]: Will add this.

**Pull request merged**